# Spam Detector

In [59]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [60]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

Given the task of creating and comparing a Logistic Regression model and a Random Forest Classifier on the provided spam detection dataset, I can make predictions about their expected performance based on the nature of the data and the inherent characteristics of each model.

Logistic Regression

- Logistic Regression is a statistical method for predicting binary outcomes based on independent variables. It is particularly well-suited for problems where the relationship between the independent variables and the log odds of the dependent variable is linear. In the context of spam detection, Logistic Regression could effectively identify relationships between the presence of specific words or characters (features) and the likelihood of an email being spam. However, its performance might be limited if the decision boundary is complex or non-linear.

Random Forest Classifier

- Random Forest, an ensemble method involving multiple decision trees, is known for its high accuracy, robustness to noise, and ability to handle non-linear data. It works well with complex datasets by capturing non-linear relationships between features and the target variable. Given the high dimensionality and potential complexity of relationships in spam detection data, Random Forest could be particularly adept at classifying emails accurately by considering various combinations of features and their interactions.

Prediction

- Based on these considerations, my prediction is that the Random Forest Classifier will outperform the Logistic Regression model in this spam detection task. This expectation stems from the Random Forest's superior ability to handle complex and non-linear relationships between features, which is common in text data like emails. The ensemble approach of Random Forest can also mitigate overfitting—a potential issue with high-dimensional data—thereby improving its generalization to unseen data.

- Moreover, while Logistic Regression offers the advantage of simplicity and interpretability, it may not capture the intricacies of the data as effectively as Random Forest. Therefore, my guess leans towards the Random Forest Classifier for achieving higher accuracy in this context, given its robustness and flexibility in dealing with various data structures and relationships.

## Split the Data into Training and Testing Sets

In [61]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.drop('spam', axis=1)

In [62]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data['spam'].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [63]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
2766             0.0               0.00           0.00           0.0   
1957             0.0               0.00           0.48           0.0   
1507             0.0               0.17           0.00           0.0   
909              0.0               1.96           0.98           0.0   
4176             0.0               0.00           0.66           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
2766            0.0            0.00              0.00                0.00   
1957            0.0            0.00              0.00                0.00   
1507            0.0            0.00              0.17                0.52   
909             0.0            0.00              1.96                0.00   
4176            0.0            0.66              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
2766              0.0            0.00  ...                   0.0        0.000   
1957              0.0            0.00  ...                   0.0        0.000   
1507              0.0            0.17  ...                   0.0        0.029   
909               0.0            0.00  ...                   0.0        0.000   
4176              0.0            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
2766        0.000        0.000        0.000        0.000        0.000   
1957        0.000        0.000        0.000        0.000        0.000   
1507        0.147        0.029        0.117        0.058        0.235   
909         0.168        0.000        1.011        0.000        0.000   
4176        0.000        0.000        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
2766                       1.600                           4   
1957                       1.353                           7   
1507                       3.521                          39   
909                        2.888                          12   
4176                       1.142                           4   

      capital_run_length_total  
2766                         8  
1957                        88  
1507                       419  
909                         52  
4176                        56  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [64]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [65]:
# Fit the Standard Scaler with the training data
scaler.fit(X_train)

StandardScaler()

In [66]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [67]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression(random_state=1)
logistic_model.fit(X_train_scaled, y_train)


LogisticRegression(random_state=1)

In [68]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_model.predict(X_test_scaled)
# Review the predictions
testing_predictions

array([0, 1, 1, ..., 1, 1, 0], dtype=int64)

In [69]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
logistic_accuracy = accuracy_score(y_test, testing_predictions)
logistic_accuracy

0.9322328410078193

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [70]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
random_forest_model = RandomForestClassifier(random_state=1)
random_forest_model.fit(X_train_scaled, y_train)

RandomForestClassifier(random_state=1)

In [71]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_testing_predictions = random_forest_model.predict(X_test_scaled)

# Review the predictions
rf_testing_predictions 

array([0, 1, 1, ..., 1, 1, 0], dtype=int64)

In [72]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_accuracy = accuracy_score(y_test, rf_testing_predictions)
rf_accuracy

0.9530842745438749

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forest Classifier model performed better with an accuracy score of approximately 95.31%, compared to the Logistic Regression model, which had an accuracy score of approximately 93.22%.

This outcome still aligns with my earlier prediction where I anticipated that the Random Forest Classifier would perform better due to its ability to model complex relationships and interactions between features, which is typical in spam detection tasks. The ensemble method of Random Forests tends to provide more robustness and is generally less prone to overfitting compared to Logistic Regression, which can be particularly advantageous in datasets with a high dimensionality.

The observed performance, even with the updated scores, continues to confirm the hypothesis that the non-linear decision boundaries and interactions between features in spam detection are better captured by the Random Forest algorithm.